staging/src/k8s.io/apimachinery/pkg/runtime/converter.go

```diff
@@ -358,41 +388,64 @@ func pointerFromUnstructured(sv, dv reflect.Value) error {
 	dv.Set(reflect.New(dt.Elem()))
 	switch st.Kind() {
 	case reflect.Ptr, reflect.Interface:
-		return fromUnstructured(sv.Elem(), dv.Elem())
+		return c.fromUnstructured(sv.Elem(), dv.Elem(), fields)
 	default:
-		return fromUnstructured(sv, dv.Elem())
+		return c.fromUnstructured(sv, dv.Elem(), fields)
 	}
 }
 
-func structFromUnstructured(sv, dv reflect.Value) error {
+func (c *unstructuredConverter) structFromUnstructured(sv, dv reflect.Value, fields map[string]struct{}) error {
 	st, dt := sv.Type(), dv.Type()
 	if st.Kind() != reflect.Map {
 		return fmt.Errorf("cannot restore struct from: %v", st.Kind())
 	}
+	keys := sv.MapKeys()
 
 	for i := 0; i < dt.NumField(); i++ {
 		fieldInfo := fieldInfoFromField(dt, i)
 		fv := dv.Field(i)
-
 		if len(fieldInfo.name) == 0 {
 			// This field is inlined.
-			if err := fromUnstructured(sv, fv); err != nil {
+			if err := c.fromUnstructured(sv, fv, fields); err != nil {
 				return err
 			}
 		} else {
+			// delete from the source's keys
+			deleteFromKeys(fieldInfo.name, &keys)
+			delete(fields, fieldInfo.name)
```

Comment:
It's a little sad that we have to change all the function signatures to carry this field that is used here. Let's see if we can do better later.